File: backend/src/main/java/com/apocalypse/thefall/entity/character/stats/SkillInstance.java
package com.apocalypse.thefall.entity.character.stats;

import com.apocalypse.thefall.entity.character.Character;
import com.apocalypse.thefall.entity.common.BaseEntity;
import jakarta.persistence.*;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

@Entity
@Getter
@Setter
@SuperBuilder
@NoArgsConstructor
@Table(name = "skill_instance", uniqueConstraints = @UniqueConstraint(columnNames = {"character_id", "skill_id"}))
public class SkillInstance extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "character_id", nullable = false)
    private Character character;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "skill_id", nullable = false)
    private Skill skill;

    @Column(name = "value", nullable = false)
    private Integer value;

    @Column(name = "is_tagged", nullable = false)
    @Builder.Default
    private boolean tagged = false;
}

File: backend/src/main/java/com/apocalypse/thefall/entity/character/stats/SkillInstance.kt
package com.apocalypse.thefall.entity.character.stats

import com.apocalypse.thefall.entity.character.Character
import com.apocalypse.thefall.entity.character.stats.enums.SkillEnum
import com.apocalypse.thefall.entity.common.BaseEntity
import com.apocalypse.thefall.service.character.rules.CalculatedInstance
import jakarta.persistence.*

@Entity
@Table(
    name = "skill_instance",
    uniqueConstraints = [UniqueConstraint(columnNames = ["character_id", "skill_id"])]
)
open class SkillInstance : BaseEntity(), CalculatedInstance<SkillEnum> {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "character_id", nullable = false)
    open var character: Character? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "skill_id", nullable = false)
    open var skill: Skill? = null

    @Column(name = "value", nullable = false)
    override var value: Int = 0

    @Transient
    override var calculatedValue: Int? = null

    @Column(name = "is_tagged", nullable = false)
    open var tagged: Boolean = false

    override val code: SkillEnum get() = skill!!.code
}
